solrmonitor/solrmonitor.go

```diff
@@ -312,33 +312,34 @@ func (c *SolrMonitor) shouldWatchChildren(path string) bool {
 
 func (c *SolrMonitor) dataChanged(path string, data string, version int32) error {
 	collectionsPrefix := c.solrRoot + "/collections/"
-	if strings.HasPrefix(path, collectionsPrefix) {
-		// we will always need the collection
-		coll := c.getCollFromPath(path)
-		if strings.HasSuffix(path, "/state.json") {
-			// common state.json change
-			if coll != nil {
-				state := coll.setStateData(data, version)
-				c.callSolrListener(coll.name, state)
-				if coll.isPRSEnabled() {
-					coll.startMonitoringReplicaStatus()
-				}
-			}
-			return nil
-		} else if trimmed := strings.TrimPrefix(path, collectionsPrefix); !strings.Contains(trimmed, "/") {
-			// less common (usually just initialization) collection change
-			if coll != nil && coll.name == trimmed {
-				state := coll.setCollectionData(data)
-				if state != nil {
-					// for collection data, we don't hit the callback until the collection state is available as well
-					// on delete, the standard state.json flow will handle passing back the nil state once cleared
-					c.callSolrListener(coll.name, state)
-				}
-			}
-			return nil
+	if !strings.HasPrefix(path, collectionsPrefix) {
+		// Expecting a collection in the /collections/ folder
+		return fmt.Errorf("unknown dataChanged: %s", path)
+	}
+
+	coll := c.getCollFromPath(path)
+	if coll == nil {
+		// Always require a collection!
+		return fmt.Errorf("unknown dataChanged: %s", path)
```

Comment:
Only concern is here. I feel like there might have been valid cases where a collection doesn't exist but we receive a callback still -- maybe on deletion or something? I would just be worried this would spur errors in certain cases.

Reply:
Can we codify that case and the expected behavior? Should we just return nil here?

Reply:
We still need to return. I guess the only question is if we return an error.

Reply:
Yeah, let me see if I can come up with a test case or something that can show the behavior. If I can't, then we can keep this! I may be totally misremembering.

Reply:
You know, looking at the code I think this is fine regardless. A new watch is set up even if an error is returned from the callback. I'm going to take one last pass and then merge this into my branch! Thanks for the cleanup!

Reply:
Alright, I at least added a delete to the tests to ensure that no error is logged in the basic case. These updates LGTM!